### DISCUSSION TITLE
Adding a compile time check without impacting existing apps compile hooks

### DISCUSSION THREAD

**Discussion by AbdealiLoKo** [sqlalchemy/sqlalchemy#10367] [Usage Questions]
Hi, I am working on a library, and wanted to add a compile time check for some things.

Mainly, I am trying to make sqla usage consistent across DBs.
For example - if I currently use `Text()` in sqlalchemy. And there is a binary expression `Text() == 'a'`
Currently, when I run this against a sqlite DB - it works. When I run it against an Oracle DB - it fails.

What I want is for a user to get an error or warning that "Oracle does not support == with TEXT" even if they are using sqlite or MySQL. So that a user does not have to get a real Oracle DB server to test their code with oracle.

I currently use the following code:
```py
@compiles(BinaryExpression)
def compile_binary_expr(binary, compiler, override_operator=None, **kw):
    """ Oracle CLOB does not support == and != operators. So, throw an error """
    operator = binary.operator
    if operator in (eq, ne, gt, ge, lt, le, in_op, notin_op):
        if isinstance(binary.left, ColumnClause) and isinstance(binary.left.type, Text):
            raise AssertionError(f'Comparison operator "{operator}" should not be used for Text column "{text_col}" as it is not supported in Oracle.")
    return compiler.visit_binary(binary, override_operator=override_operator, **kw)
```

But the issue I am realizing is that if I do this - the user will not be able to write their own `compiles(BinaryExpression)` check.
Because I can override the `BinaryExpression` compilation only once. (Which makes sense)
And many of my applications are already overriding this compilation, so I did not want to impose this constraint that my `compiles` will limit how they compile stuff.

Any recommendations on how I can achieve this ? Something like "pre compile check" ?

**Comment by zzzeek** [comment]
> But the issue I am realizing is that if I do this - the user will not be able to write their own compiles(BinaryExpression) check.

that's correct, we dont have support for bundling multiple compile checks like this right now.

> Any recommendations on how I can achieve this ? Something like "pre compile check" ?

if this is a library I'd probably want to stay away from any global `@compiles` hook like that, IMO libraries should not do "global" things.

it seems like you are intending to make some kind of custom dialect, like "Oracle on MySQL" or something like that ?  I think you would want to provide your own SQLCompiler class that subclasses the existing SQLCompiler for each dialect you want to target, and it gets installed on a per-dialect/engine basis.   this "plugging in" of an alternate compiler can be done cleanly using the https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.CreateEnginePlugin .

**Comment by CaselIT** [comment]
> And many of my applications are already overriding this compilation

This seems surprising. What's the use case?
(sorry it's a bit off topic, just wanted to ask in case there is an api missing in sqlalchemy)

**Reply by zzzeek** [reply]
well missing for us is that you can't put mutliple `@compiles` hooks on something

**Reply by CaselIT** [reply]
well yes, but I can't think of a reason for overriding the compilation of `BinaryExpression` in many cases.

@zzzeek maybe to support this feature `compiler` we could defined a new function so that you could do `compiler.next(thing, **kw)` and it will call the next `compiles` registered or the default version if all were already called. I guess you would need a way of ordering the multiple compiles at that point though

**Reply by AbdealiLoKo** [reply]
I actually rely on `compiles()` a fair bit to handle weird quirks in the DBs

My application at least has 2-3 patches on Binary Expression

For example, I have a patch for Oracle's limit of having a maximum of 1000 items for `IN` clause
It takes in `id IN (1,2,3,4,5,...2000)` and breaks it down into `id IN (1,2, ..., 1000) OR id IN (1001, 1002, ..., 2000)` 

In another case, postgres does not support auto casting an integer with boolean in the binary expression.
So, if someone gives `WHERE col1 == 1` and col1 is Boolean - it throws an error saying postgres does not support this.

Another case is mssql does not support `IS FALSE` and needs `== FALSE`

Currently what I have to do is:
```
@compiles(BinaryExpression)
def compile_binary_expr(binary, compiler, override_operator=None, **kw):
    # Handle case 1 and continue
    # Handle case 2 and continue
    # Handle case 3 and continue
    # Else
    return compiler.visit_binary(binary, override_operator=override_operator, **kw)
```
I wanted to move some of these cases into a library. And I can't do that currently

NOTE - I run my compile checks on all DBs by default. Because if I want to give warnings/errors to users saying that for a specific DB (even if that is not the DB that was connected to) their code may break

**Reply by AbdealiLoKo** [reply]
I don't think that having multiple `@compiles` is a good idea because `@compiles` is a "transformation"
And allowing multiple implies you need to have an order of transformations and the ability to chain them. Which I think @CaselIT was referring to.
In such cases as a library author, I would rather just give my transformation logic as a function and get my user to write the `@compiles` themselves in the right order

So, I think that the missing part is something like a `@pre_compile` which does not return anything. But allows me to just run a function before the `@compiles` runs.

**Reply by CaselIT** [reply]
undestood.

> For example, I have a patch for Oracle's limit of having a maximum of 1000 items for IN clause
It takes in id IN (1,2,3,4,5,...2000) and breaks it down into id IN (1,2, ..., 1000) OR id IN (1001, 1002, ..., 2000)

note that this is not cache safe on sqlalchemy 1.4+

this actually seems like something sqlalchemy could do automatically. What do you think @zzzeek ?

> So, I think that the missing part is something like a @pre_compile which does not return anything. But allows me to just run a function before the @compiles runs.

There is something like this called form linters, that is what raises the warning when compiling a statement with a cartesian product. I'm not sure how public they are and if they work also for non form things though. Since it's not documented it's likely not public, should it @zzzeek?
The code is here https://github.com/sqlalchemy/sqlalchemy/blob/83fa1d92ef148bf19bab5e716537b13a08c7574e/lib/sqlalchemy/sql/compiler.py#L673

**Reply by zzzeek** [reply]
I'd rather not get into automatically scaling IN phrases like that, that's too much magic / surprise

**Reply by CaselIT** [reply]
> I'd rather not get into automatically scaling IN phrases like that, that's too much magic / surprise

yes, this is probably similar to transforming `in` `not in` into `any` `all` on pg I guess.
probably not something worth doing at the moment.

**Comment by zzzeek** [comment]
> well yes, but I can't think of a reason for overriding the compilation of `BinaryExpression` in many cases.
> 
> @zzzeek maybe to support this feature `compiler` we could defined a new function so that you could do `compiler.next(thing, **kw)` and it will call the next `compiles` registered or the default version if all were already called. I guess you would need a way of ordering the multiple compiles at that point though

yeah it's a weird case because each `@compiles` would have to be written to support this, and what does that really mean when two different systems have a different string for something.    Issue #2814 exists for this but I dont know that we will ever do it.   the use cases here would need to be better identified and perhaps a more integrated system in the base compiler would need to be devised.

**Reply by CaselIT** [reply]
Well my suggestion there was that a compiler function world either compile something, creating the string, or call the next one. At all times only a single compile function would ever generate the sql string.

A bit like pytest hooks.

I guess we could even do that by returning a special variable from the function, but it would complicate the change of the thing or of the kwargs
